tmva/tmva/inc/TMVA/RBDT.hxx

```diff
@@ -19,108 +20,70 @@
-   }
+   /// Construct backends from model in ROOT file.
+   RBDT(const std::string &key, const std::string &filename);
 
-   /// Compute model prediction on a single event
+   /// Compute model prediction on a single event.
    ///
    /// The method is intended to be used with std::vectors-like containers,
    /// for example RVecs.
    template <typename Vector>
-   Vector Compute(const Vector &x)
+   Vector Compute(const Vector &x) const
    {
-      Vector y;
-      y.resize(fNumOutputs);
-      for (int i = 0; i < fNumOutputs; i++)
-         fBackends[i].Inference(&x[0], 1, true, &y[i]);
-      if (fNormalizeOutputs) {
-         Value_t s = 0.0;
-         for (int i = 0; i < fNumOutputs; i++)
-            s += y[i];
-         for (int i = 0; i < fNumOutputs; i++)
-            y[i] /= s;
-      }
+      std::size_t nOut = baseResponses_.size() > 2 ? baseResponses_.size() : 1;
+      Vector y(nOut);
+      ComputeImpl(x.data(), y.data());
       return y;
    }
 
-   /// Compute model prediction on a single event
-   std::vector<Value_t> Compute(const std::vector<Value_t> &x) { return this->Compute<std::vector<Value_t>>(x); }
+   /// Compute model prediction on a single event.
+   inline std::vector<Value_t> Compute(std::vector<Value_t> const &x) const { return Compute<std::vector<Value_t>>(x); }
 
-   /// Compute model prediction on input RTensor
-   RTensor<Value_t> Compute(const RTensor<Value_t> &x)
-   {
-      const auto rows = x.GetShape()[0];
-      RTensor<Value_t> y({rows, static_cast<std::size_t>(fNumOutputs)}, MemoryLayout::ColumnMajor);
-      const bool layout = x.GetMemoryLayout() == MemoryLayout::ColumnMajor ? false : true;
-      for (int i = 0; i < fNumOutputs; i++)
-         fBackends[i].Inference(x.GetData(), rows, layout, &y(0, i));
-      if (fNormalizeOutputs) {
-         Value_t s;
-         for (int i = 0; i < static_cast<int>(rows); i++) {
-            s = 0.0;
-            for (int j = 0; j < fNumOutputs; j++)
-               s += y(i, j);
-            for (int j = 0; j < fNumOutputs; j++)
-               y(i, j) /= s;
-         }
-      }
-      return y;
-   }
-};
+   RTensor<Value_t> Compute(RTensor<Value_t> const &x) const;
 
-extern template class TMVA::Experimental::RBDT<TMVA::Experimental::BranchlessForest<float>>;
-extern template class TMVA::Experimental::RBDT<TMVA::Experimental::BranchlessJittedForest<float>>;
+   static RBDT LoadText(std::string const &txtpath, std::vector<std::string> &features, int nClasses = 2);
+   static RBDT LoadText(std::istream &is, std::vector<std::string> &features, int nClasses = 2);
+
+   std::vector<int> rootIndices_;
+   std::vector<unsigned int> cutIndices_;
+   std::vector<Value_t> cutValues_;
+   std::vector<int> leftIndices_;
+   std::vector<int> rightIndices_;
+   std::vector<Value_t> responses_;
+   std::vector<int> treeNumbers_;
+   std::vector<Value_t> baseResponses_;
+   Value_t baseScore_ = 0.0;
+   bool logistic_ = false;
```

Comment:
Since I originally followed the CMSSW code style, I used trailing underscores to mark public data members. Coincidentally, the xgboost python package follows almost the same convention (only with `snake_case_`, as one can see in examples like `xgb_model.n_classes_`).

The `fDataMember` convention is for private data members AFAIK. Anyway the best solution is probably to make all these data members private anyway, and then we can go for `fStyle`.

Reply:
I have made all the data members private and use the `fDataMember` style in my latest update!